### PR TITLE
extmod/vfs_posix_file: Skip flush of tty handles in msvc debug builds.

### PR DESCRIPTION
**Pull request by stinos**
Follow up on the comments in #12576.

So in debug builds with debug error reporting set to showing a dialog (to allow attaching the debugger), any application which imports the logging module and leaves the default handlers would result in this dialog because logging.shutdown is called at exit and that flushes the default handler which has stderr as its stream. Confusing/annoying enough to fix this.

I'm not completely sure vfs_posix_file would ever be used with a serial port, but it seems better to be very explicit about what is being done here anyway.

### CODE REVIEW

File: extmod/vfs_posix_file.c [hunk 1]
```diff
@@ -160,6 +160,17 @@ STATIC mp_uint_t vfs_posix_file_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_
...
             #define VFS_POSIX_STREAM_STDIO_ERR_CATCH (err == EINVAL || err == ENOTSUP)
             #elif defined(_MSC_VER)
+            // In debug builds _commit will generate a debug report via _ASSERTE when called with
+            // non-redirected stdin/stdout/stderr handles because FlushFileBuffers, which it calls
+            // internally, will fail since console output is not buffered. In release builds this also
+            // fails, but merely returns an error which is handled appropriately below.
+            // The check for the handle being stdin/stdout/stderr is added explicitly because according to
+            // the documentation _isatty is also true for serial ports for instance.
+            #ifdef _DEBUG
+            if ((o->fd == STDIN_FILENO || o->fd == STDOUT_FILENO || o->fd == STDERR_FILENO) && _isatty(o->fd)) {
```
Comment (dpgeorge) [inline extmod/vfs_posix_file.c:170]: We could just have this check done on all platforms.  Then remove the `VFS_POSIX_STREAM_STDIO_ERR_CATCH` check below.  That would simplify the code and unify all platforms.

Comment (dpgeorge) [reply extmod/vfs_posix_file.c:170]: Unfortunately this doesn't fully work on Linux: when running the test suite it uses `subprocess.check_output`, that's not a tty so it attempts the flush syscall, and fails with EINVAL (for the `io/file_stdio2.py` test).  Not sure how to get around that.

Comment (stinos) [reply extmod/vfs_posix_file.c:170]: Might be better to leave as-is. However, I just found out that comparing against STDIN_FILENO etc is wrong, manual for `close`:

> To deallocate means to make the file descriptor available for return by subsequent calls to [open()]

so `sys.stdout.close()` followed by `open('somefile')` means the `fd` of that file is 1 i.e. same as stdout.

Comment (stinos) [reply extmod/vfs_posix_file.c:170]: Hmm, actually would this not be fixed by instead of comparing `o->fd`, comparing `o` itself against `mp_sys_stdin_obj` and the likes?

Comment (dpgeorge) [reply extmod/vfs_posix_file.c:170]: It would.  But what about the case of creating an additional stdio file instance, eg `open(1)`?

Comment (stinos) [reply extmod/vfs_posix_file.c:170]: That is now implemented as

```
if (mp_obj_is_small_int(fid)) {
    o->fd = MP_OBJ_SMALL_INT_VALUE(fid);
    return MP_OBJ_FROM_PTR(o);
}
```

so could be changed into something like

```
if (mp_obj_is_small_int(fid)) {
    int fd = MP_OBJ_SMALL_INT_VALUE(fid);
    if (fd == STDIN_FILENO) {
      return MP_OBJ_FROM_PTR(&mp_sys_stdin_obj);
    }
    //etc
}
```

But that again runs into the same problem: first sys.stderr.close() followed by opening a file one would then result in open(2) to stil return the stderr stream which is somewhat inconsistent. Plus this function doesn't actually 'open' anything in case of std file descriptors, but that is another issue.

Note that CPython seems to do things in a different way:
```
>>> import sys
>>> sys.stderr.close()
>>> sys.stderr.fileno()
2
>>> w=open('c:/temp/fo.txt', 'w')
>>> w.fileno()
3
```
it's as if it doesn't actually close the underlying stream but instead sets a flag so that it looks as if it's closed. Or perhaps there's a whole extra layer on top.

tldr; markedly complicated, any fix is going to be a slight code size increase, a 'complete' fix would be quite a lot of changes, and all of that which is in practice almost never an issue (I think?). How about no fix but just adding a comment that explains things a bit?

Comment (dpgeorge) [reply extmod/vfs_posix_file.c:170]: > How about no fix but just adding a comment that explains things a bit?

Yes, that sounds fine to me.

Comment (stinos) [reply extmod/vfs_posix_file.c:170]: Ok, added a comment to the common code below (so it's clear this isn't just for msvc)

**Review by dpgeorge — APPROVED** [review]

File: extmod/vfs_posix_file.c (hunk 1 above)
Comment (dpgeorge) [reply extmod/vfs_posix_file.c:170]: Great, thanks!
